widgetry/src/widgets/panel.rs

```diff
@@ -33,7 +34,87 @@ impl Panel {
         }
     }
 
+    fn update_container_dims_for_canvas_dims(&mut self, canvas_dims: ScreenDims) {
+        let new_container_dims = match self.dims {
+            Dims::MaxPercent(w, h) => ScreenDims::new(
+                self.contents_dims.width.min(w.inner() * canvas_dims.width),
+                self.contents_dims
+                    .height
+                    .min(h.inner() * canvas_dims.height),
+            ),
+            Dims::ExactPercent(w, h) => {
+                ScreenDims::new(w * canvas_dims.width, h * canvas_dims.height)
+            }
+        };
+        self.container_dims = new_container_dims;
+    }
+
+    // TODO: this method potentially gets called multiple times in a render pass as an
+    // optimization, we could replace all the current call sites with a "dirty" flag, e.g.
+    // `set_needs_layout()` and then call `layout_if_needed()` once at the last possible moment
     fn recompute_layout(&mut self, ctx: &EventCtx, recompute_bg: bool) {
+        let old_scrollable_x = self.scrollable_x;
+        let old_scrollable_y = self.scrollable_y;
+
+        self.scrollable_x = self.contents_dims.width > self.container_dims.width;
+        self.scrollable_y = self.contents_dims.height > self.container_dims.height;
+
+        // Unwrap the main widget from any scrollable containers if necessary.
+        if old_scrollable_y && !self.scrollable_y {
+            let container = self.top_level.widget.downcast_mut::<Container>().unwrap();
+            self.top_level = container.members.remove(0);
+        }
+
+        if old_scrollable_x && !self.scrollable_x {
+            let container = self.top_level.widget.downcast_mut::<Container>().unwrap();
+            self.top_level = container.members.remove(0);
+        }
+
+        let top_left = ctx
+            .canvas
+            .align_window(self.container_dims, self.horiz, self.vert);
+
+        // Wrap the main widget in scrollable containers if necessary.
+        if self.scrollable_x && !old_scrollable_x {
+            let mut place_holder = Widget::nothing();
+            std::mem::swap(&mut place_holder, &mut self.top_level);
```

Comment:
My Rust-fu is weak. Is there a better way to do this shuffling?

The old `self.top_level` is owned by the new `self.top_level` so there's kind of an ownership circle without temporarily introducing this `Widget::nothing()`

Reply:
You can do `let placeholder = std::mem::replace(&mut self.top_level, Widget::nothing())`. That's a slightly shorter form, equivalent to what you have here

Reply:
That is a little better - updated!